### PR TITLE
Report API rework

### PR DESCRIPTION
**Pull request by Buroni**
## Prerequisites

- [x] Has been tested locally
- [ ] If a bugfix, have included a breaking test if possible
- [ ] If UI change, have added screenshots

## Proposed Changes

- Remove report publishing logic from the `Report` class into `Builder`, `Server`, and `Saver` classes which take a `Report` instance
- Rename `Report` to `App` in user-facing identifiers and comments. Internal methods still use "`report`" as we're trying not to affect the entire codebase with this initial refactor
- Rename `App` to `LegacyApp`
- Add deprecation warnings to `Report` constructor, as well as `build` and `save` report methods

## Notes

- I've disabled the legacy app tests in `client.e2e.test_apps` -- confirming do we want to disable these seeing as apps are retired in 0.15+?
- We're not updating docs or READMEs n this PR following discussion with @mands 
- I'm using `warnings.warn` to show the `DeprecationWarning`, however afaik this is only shows when enabled by the user e.g. via `warnings.simplefilter`. Is there a better way to display `DeprecationWarning`s?

## TODO

- ~Fix failing Ubuntu tests~

### CODE REVIEW

File: projects/python-client/tests/client/e2e/test_apps.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 # TODO: FIX TEST TO USE ENVIRONMENT VALUE FROM ENVIRONMENT OBJECT
 
 
+@pytest.mark.skip
```
Comment (mands) [inline projects/python-client/tests/client/e2e/test_apps.py:18]: you can do this at the entire module level - see https://docs.pytest.org/en/latest/how-to/skipping.html#summary

```python
# at top of module before all tests
pytestmark = pytest.mark.skip("LeagacyApp tests disabled")
```

File: projects/python-client/src/datapane/client/api/report/core.py [hunk 2]
```diff
@@ -60,21 +61,28 @@ def end_headers(self):
...
-class ReportWidth(Enum):
+class AppWidth(Enum):
     """The document width"""
 
     NARROW = "narrow"
     MEDIUM = "medium"
     FULL = "full"
 
 
+class ReportWidth(Enum):
```
Comment (mands) [inline projects/python-client/src/datapane/client/api/report/core.py:79]: could this just inherit from `AppWidth`?

Comment (mands) [reply projects/python-client/src/datapane/client/api/report/core.py:79]: or even an alias, i.e. `ReportWidth = AppWidth`?

File: projects/python-client/src/datapane/client/api/report/core.py [hunk 3]
```diff
@@ -127,14 +140,14 @@ def to_css(self) -> str:
...
-    # Escape </script> to prevent 3rd party JS terminating the local report bundle.
+    # Escape </script> to prevent 3rd party JS terminating the local app bundle.
     # Note there's an extra "\" because it needs to be escaped at both the python and JS level
     src = env.loader.get_source(env, name)[0].replace("</script>", r"<\\/script>")
     return Markup(src)
 
 
-class ReportFileWriter:
-    """Provides shared logic for saved and served local report writers"""
+class AppFileWriter:
```
Comment (mands) [inline projects/python-client/src/datapane/client/api/report/core.py:149]: is this needed as a standalone class or can it be moved into the Processor base?

File: projects/python-client/src/datapane/client/api/report/core.py [hunk 4]
```diff
@@ -242,17 +253,54 @@ def __init__(
...
+        tags: t.List[str] = None,
+        project: t.Optional[str] = None,
+        open: bool = False,
+        formatting: t.Optional[AppFormatting] = None,
+        overwrite: bool = False,
+        **kwargs,
+    ) -> None:
+        warn(
+            "`App.upload` is deprecated and will be removed from future releases, use `Uploader` instead",
+            DeprecationWarning,
```
Comment (mands) [inline projects/python-client/src/datapane/client/api/report/core.py:282]: for `DeprecationWarning`, I think we also set `stacklevel=2` so the warning refers to where the user called the deprecated code (see https://dev.to/hckjck/python-deprecation-2mof)

File: projects/python-client/src/datapane/client/api/report/core.py [hunk 5]
```diff
@@ -265,6 +313,36 @@ def _preprocess_pages(self, pages: t.List[BlockOrPrimitive]):
...
+        **kwargs,
+    ):
+        warn(
+            "`Report` is deprecated and will be removed from future releases, use `App` instead",
+            DeprecationWarning,
+        )
+        super().__init__(*arg_blocks, blocks=blocks, layout=layout, **kwargs)
+
+
+class Processor:
```
Comment (mands) [inline projects/python-client/src/datapane/client/api/report/core.py:336]: I think we can move `Processor` and all its implementations into a new `processors` module

Comment (mands) [reply projects/python-client/src/datapane/client/api/report/core.py:336]: also, let's mark as an abstractclass

**Review by mands — COMMENTED** [review]
Looks good - I think a few small refactorings and can merge

File: projects/python-client/src/datapane/client/api/report/core.py (hunk 3 above)
Comment (Buroni) [reply projects/python-client/src/datapane/client/api/report/core.py:149]: Yeah it makes sense to move the writer logic into the `Processor` hierarchy. I've added it to an intermediate `LocalProcessor` as only the `Saver` and `Server` classes need to use it.

File: projects/python-client/src/datapane/client/api/report/core.py (hunk 5 above)
Comment (Buroni) [reply projects/python-client/src/datapane/client/api/report/core.py:336]: Sounds good re moving into own module. 

> also, let's mark as an abstractclass

Reports are wrapped in `Processor` to generate report documents for testing, e.g. in `tests.client.local.api.test_reports` we do `dp.Processor(r)._gen_report`

Comment (mands) [reply projects/python-client/src/datapane/client/api/report/core.py:336]: Oh ok, yeah that's fine then

**Comment by Buroni** [comment]
Have pushed feedback

I fixed a bug where custom report width isn't being applied to served and local reports. This required the dictionary `widths` defined in `dp_server/.../dp_tags` -- I've copied this dict over to the client for now with a `TODO`, as moving the dict into `datapane.common.report` created some circular dependency issues 

Also the Ubuntu tests are failing on `datapane.client.utils.InvalidTokenError`. @mands do you know why those tests in particular might be failing?

**Comment by mands** [comment]
Sounds good re the changes - unsure re the error, will take a look this afternoon

**Comment by mands** [comment]
This is good to merge - all tests passing now, just waiting on docs (which I've updated to only run on non-draft PRs)

**Pull request merged**